### PR TITLE
[New Feature][Habana-Main] speculative_decoding HPU support

### PR DESCRIPTION
**Pull request by xuechendi**
Req - https://jira.habana-labs.com/browse/REQ-289 => target for 1.19

TODO:
- There remains one hardcode to HPUWorker, need to remove

Next Steps:

- 1. submit necessary codes change to vllm-upstream branch => WIP
- 2. support all 3 draft_model_types - mlp_speculator, medusa and others

FILL IN THE PR DESCRIPTION HERE

FIX #xxxx (*link existing issues this PR will resolve*)

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by xuechendi** [comment]
Run test with 
```
python examples/offline_inference_spec_decode.py
```

image

**Review by michalkuligowski — CHANGES_REQUESTED** [review]

File: vllm/worker/hpu_model_runner.py [hunk 1]
```diff
@@ -338,6 +338,10 @@ def compute_logits(self, *args, **kwargs):
     def sample(self, *args, **kwargs):
         return self.model.sample(*args, **kwargs)
 
+    @property
+    def sampler(self):
```
[the remark this answers is not in the document]
Comment (xuechendi) [reply vllm/worker/hpu_model_runner.py:342]: @michalkuligowski , 'sampler' will be invoked in spec_decode_worker, rename to 'model_sampler' will bring lots of codes change.

I paste an examples as below
image

File: vllm/spec_decode/metrics.py [hunk 2]
```diff
@@ -77,13 +78,20 @@ def __init__(self,
...
+            import habana_frameworks.torch as htorch
+            self._copy_stream = htorch.hpu.Stream()
+        else:
+            self._copy_stream = torch.cuda.Stream()
 
     def maybe_collect_rejsample_metrics(
             self, k: int) -> Optional[SpecDecodeWorkerMetrics]:
 
+        if not current_platform.is_cuda_alike():
+            return None
```
[the remark this answers is not in the document]
Comment (xuechendi) [reply vllm/spec_decode/metrics.py:93]: @michalkuligowski , for non_cuda backend, this function is not usable because it hard code with `torch.cuda.Event`
=> This will not affect workflow, so I think we might skip it to make functionality ready?

I provide an example as below
```
    def _collect_rejsample_metrics(
            self, k: int,
            ready_event: torch.cuda.Event) -> SpecDecodeWorkerMetrics:
```

```
def _copy_rejsample_metrics_async(self) -> torch.cuda.Event:
        """Copy rejection/typical-acceptance sampling metrics
        (number of accepted tokens, etc) to CPU asynchronously.

        Returns a CUDA event recording when the copy is complete.
        """
        assert self._copy_stream is not None
        self._copy_stream.wait_stream(torch.cuda.current_stream())

        with torch.cuda.stream(self._copy_stream):
            self._aggregate_num_accepted_tokens.copy_(
                self.spec_decode_sampler.num_accepted_tokens,
                non_blocking=True)
            self._aggregate_num_emitted_tokens.copy_(
                self.spec_decode_sampler.num_emitted_tokens, non_blocking=True)
            # Number of draft tokens is calculated on CPU, so no copy is
            # required.
            self._aggregate_num_draft_tokens = (
                self.spec_decode_sampler.num_draft_tokens)

        aggregate_metrics_ready = torch.cuda.Event()
        aggregate_metrics_ready.record(self._copy_stream)

        return aggregate_metrics_ready
```

File: vllm/worker/selector.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+from vllm.worker.hpu_worker import HPUWorker
+
+def init_worker(*args, **kwargs):
+    device_config: DeviceConfig = kwargs.get("device_config")
+    if device_config.device_type == 'cuda':
+        return Worker(*args, **kwargs)
+    elif device_config.device_type == 'hpu':
+        return HPUWorker(*args, **kwargs)
+    else:
+        raise NotImplementedError("Please help to add your preferred backend")
```
[the remark this answers is not in the document]
Comment (xuechendi) [reply vllm/worker/selector.py]: Fixed by adding all platform in elif

File: examples/offline_inference_spec_decode.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    del llm
+    gc.collect()
+    print("================= Summary =====================")
+    print("input is ", prompts)
+    print()
+    print("Non Spec Decode - latency_per_token is ", latency_per_token_non_spec)
+    print("Generated Text is :", ret_non_spec)
+    print()
+    print("Spec Decode - latency_per_token is ", latency_per_token_spec)
+    print("Generated Text is :", ret_spec)
```
[the remark this answers is not in the document]
Comment (xuechendi) [reply examples/offline_inference_spec_decode.py]: updated

**Comment by xuechendi** [comment]
@michalkuligowski , I fixed all comments

File: vllm/model_executor/layers/spec_decode_base_sampler.py [hunk 5]
```diff
@@ -77,7 +77,8 @@ def _create_output(
             tensor is [batch_size, k + num_bonus_tokens]
         """
         batch_size, k = substitute_token_ids.shape
-        bonus_token_ids = bonus_token_ids.squeeze()
+        # TODO: HPU has an issue when handling squeeze with all dim as 1
+        bonus_token_ids = bonus_token_ids.squeeze(-1)
```
Comment (michalkuligowski) [inline vllm/model_executor/layers/spec_decode_base_sampler.py:81]: shouldn't it be under if hpu clause?

File: vllm/spec_decode/draft_model_runner.py [hunk 6]
```diff
@@ -12,21 +22,10 @@
         # vllm_flash_attn is not installed, try the ROCm FA metadata
         from vllm.attention.backends.rocm_flash_attn import (
             ROCmFlashAttentionMetadata as FlashAttentionMetadata)
-except (ModuleNotFoundError, ImportError) as err:
-    raise RuntimeError(
+except Exception as e:
```
Comment (michalkuligowski) [inline vllm/spec_decode/draft_model_runner.py:25]: you are changing behavior here from terminating to continuing with warning

**Review by michalkuligowski — CHANGES_REQUESTED** [review]

File: vllm/spec_decode/draft_model_runner.py (hunk 6 above)
Comment (xuechendi) [reply vllm/spec_decode/draft_model_runner.py:25]: Hi, @michalkuligowski , yes, the spec_decode_worker.py imports draft_model_runner.py. And above ModuleNotFoundError in draft_model_runner.py will cause termination. That is why I changed the beviour to avoid termination. 

If this is not preferred, I can move the `from vllm.spec_decode.draft_model_runner import TP1DraftModelRunner` into `SpeDecoderWorker.create_worker` as below to avoid unwanted importError termination. => Will that be more acceptable to you?

```
else:
                if draft_tp == 1:
                    from vllm.spec_decode.draft_model_runner import TP1DraftModelRunner
                    draft_worker_kwargs[
                        "model_runner_cls"] = TP1DraftModelRunner
                else:
                    if draft_worker_kwargs[
                            "model_config"].hf_config.model_type == "eagle":
                        raise NotImplementedError(
                            "EAGLE does not support TP > 1 yet")
```

File: vllm/model_executor/layers/spec_decode_base_sampler.py (hunk 5 above)
Comment (xuechendi) [reply vllm/model_executor/layers/spec_decode_base_sampler.py:81]: > shouldn't it be under if hpu clause?

@michalkuligowski , I am not getting this question. This is what I observed during debug, do you want me to remove the comments?

Comment (xuechendi) [reply vllm/model_executor/layers/spec_decode_base_sampler.py:80]: remove the comments

File: vllm/spec_decode/draft_model_runner.py (hunk 6 above)
Comment (xuechendi) [reply vllm/spec_decode/draft_model_runner.py]: Fixed above comments by check with platform. and only raise when platform is not one of Cuda, Roc or HPU.

**Comment by xuechendi** [comment]
@michalkuligowski , I updated the codes according to your last comments. For the draft_model_runner.py importing behavior change issue you mentioned in last comment, since draft_model_runner will be imported by spec_decode_runner and multi_step_decode_runner, we need to prevent the unnecessary importing error termination due to Cuda and ROCm flashattn support. 
Since you think the previous fixing by simply changing raise to print is not very good, I provided an alternative fixing in latest commit.

After this PR merged, I will move on to the complete draft model support for medusa and mlp, and I'll revisit this draft_model_runner.py for better support.

File: vllm/spec_decode/draft_model_runner.py [hunk 7]
```diff
@@ -4,28 +4,22 @@
...
-                                                  FlashAttentionMetadata)
-else:
+        except:
+            from vllm.attention.backends.hpu_attn import (
+                HPUPagedAttentionMetadata as FlashAttentionMetadata)
+except (ModuleNotFoundError, ImportError, AssertionError) as err:
     raise RuntimeError(
         "Draft model speculative decoding currently only supports"
-        "CUDA and ROCm and HPU attention backend.")
+        "CUDA and ROCm and HPU attention backend.") from err
```
Comment (xuechendi) [inline vllm/spec_decode/draft_model_runner.py:23]: @michalkuligowski , the previous CI failing is caused by ruff and yapf gave me conflicted fixing suggestion. So I changed the codes as above, which is acceptable to both yapf and ruff now.

**Comment by xuechendi** [comment]
@michalkuligowski , please help to review. 

The previous commit of using platform to switch between cuda-alike and hpu leads to a coding format issue, which yapf and ruff gave me different fixing suggestion and I can't get both of them passed the check.

That is why I have to change it to use try and expect.

**Comment by xuechendi** [comment]
@michalkuligowski , rebased this PR to latest habana_main. 

As mentioned in last commit, in order to pass CICD formatting check, I have to still use "try and except" in draft_model_runner.py. Previous way of using "platform check" causes a formatting conflict issue that yapf prefers one format and isort asks for another one, so I can't make CICD passed.

Please have a review.

**Comment by xuechendi** [comment]
@michalkuligowski , I submitted my second PR: https://github.com/HabanaAI/vllm-fork/pull/461

PR461 is based on PR375, and enabled Medusa and MLP speculator there.

File: vllm/worker/selector.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
+from vllm.config import DeviceConfig
```
Comment (michalkuligowski) [inline vllm/worker/selector.py:1]: The code looks fine, please rebase and test if multistep scheduling is not broken with this PR

Comment (xuechendi) [reply vllm/worker/selector.py]: Codes rebased to latest habana_main

**Comment by xuechendi** [comment]
@michalkuligowski , I have rebased this PR to latest habana_main - changes are in [20a2e6](https://github.com/HabanaAI/vllm-fork/pull/375/commits/20a2e673f5af93ff08ec907083849019e621fb58)

Meanwhile, I tested with multi_step_scheduler using cmdline as below => passed (The multi-step-worker in spec_decode is different to the one for multi-step-scheduler => so it will not affect multi-step-scheduler)

```
VLLM_PROFILER_ENABLED=true VLLM_SKIP_WARMUP=true VLLM_CONTIGUOUS_PA=true VLLM_PA_SOFTMAX_IMPL='wsum' python benchmarks/benchmark_throughput.py \
        --model ${model} \
        --device hpu \
        --backend vllm \
        --num-prompts ${bs} \
        --input_len ${in_len} \
        --output_len ${out_len} \
        --max_model_len ${total_len} \
        --dtype bfloat16 \
        --num_scheduler_steps 16 \
        --tensor_parallel_size 1 \
        --gpu-memory-util 0.9
```

**Review by michalkuligowski — APPROVED** [review]
